File: React_Native/React_Native/src/pages/sat_pages/ativarSat.js
import React, { Component } from "react";

import { Text, View, Alert,StyleSheet,TextInput,Button, DeviceEventEmitter } from "react-native";
import { SafeAreaView } from "react-native-safe-area-context";
import {TextInputMask} from 'react-native-masked-text';

import ValidaCodigo from '../../components/ValidacaoCodigo';
import ValidacaoCnpj from "../../components/ValidacaoCnpj";

import OperacaoSat from '../../services/operacaoSat';
import RetornoSat from '../../services/retornoSat';

export default class AtivarSat extends Component{
    constructor(props) {
        super(props);       
        this.state = {
            cnpj:'',// CNPJ do contribuinte
            codigoAtivacao:global.MyVar,
            confirmacaoCodigo:''
        };
    }

    dialogoSat(messageText){
        Alert.alert("Retorno", messageText);
    }

    ativarSat(){
        // Salva  o código de ativação para o usuario não precisar ficar digitando em todas as telas
        global.MyVar = this.state.codigoAtivacao;

        // CRIA UMA INSTÂNCIAS DO COMPONENTE QUE VALIDA O CÓDIGO E CNPJ
        var validationCodigo = new ValidaCodigo();
        var validationCnpj = new ValidacaoCnpj();

        //LIMPEZA DO CNPJ DE ENTRADA
        var cnpj = validationCnpj.getClean(this.state.cnpj.toString());    

        if(validationCodigo.getValidation(this.state.codigoAtivacao)){
            if(this.state.codigoAtivacao==this.state.confirmacaoCodigo){
                if(validationCnpj.getSize(cnpj)){
                    this.dialogoSat("Verifique o CNPJ digitado!");
                }else{
                    
                    //ENVIA PARA OPERAÇÃO SAT ONDE VAI CONECTAR COM O SAT DE ACORDO COM A FUNÇÃP
                    var operacaoSat = new OperacaoSat();
                    args = {'funcao': 'AtivarSAT',
                            'random': Math.floor(Math.random() * 9999999).toString(),
                            'codigoAtivar': this.state.codigoAtivacao,
                            'cnpj': this.state.cnpj};
                    operacaoSat.invocarOperacaoSat(args);
                    
                    //RECEBE OS DADOS QUE FORAM ENVIADOS NA REPOSTA
                    DeviceEventEmitter.once("eventAtivar", event=> { 
                        var retorno = event.ativar;
                        var retornoSat = new RetornoSat(args['funcao'], retorno);
                        this.dialogoSat(operacaoSat.formataRetornoSat(retornoSat));
                    });
                }

            }else{
                this.dialogoSat("O Código de Ativação e a Confirmação do Código de Ativação não correspondem!");
            }
        }else{
            //console.log('XXXX: ', validationCnpj.getValidation(this.state.codigoAtivacao));
            this.dialogoSat("Código de Ativação deve ter entre 8 a 32 caracteres!");
        }
    }

    render() {
        return (
            <SafeAreaView>
                <View style={styles.containerCnpj}>
                    <Text style={styles.titleDefault}>CNPJ Contribuinte</Text>
                    <View style={styles.containerInput}>
                        <TextInputMask
                            type={'cnpj'}
                            value={this.state.cnpj}
                            onChangeText={text => {
                                this.setState({
                                cnpj: text
                                })
                            }}
                        />
                    </View>
                </View>
                    
                <View style={styles.containerDefault}>
                    <Text style={styles.titleDefault}>Código de Ativação Sat</Text>
                    <TextInput           
                        style={styles.inputDefault}  
                        value={this.state.codigoAtivacao}
                        onChangeText={(text) => this.setState({codigoAtivacao: text})} 
                    />
                </View>

                <View style={styles.containerDefault}>
                    <Text style={styles.titleDefault}>Confirmação do Código</Text>
                    <TextInput           
                        style={styles.inputDefault}    
                        value={this.state.confirmacaoCodigo}
                        onChangeText={(text) => this.setState({confirmacaoCodigo: text})}  
                    />
                </View>
                    
                <View style={styles.containerButton}>
                    <Button
                        color= 'gray'
                        title="ativar"
                        onPress={() => this.ativarSat()}
                        />
                </View>
                
            </SafeAreaView>    
        );
    } 
}

const styles = StyleSheet.create({
    containerCnpj: {
        marginBottom:10,
        marginTop: 20
    },
    titleDefault: {
        fontSize:20,
        fontWeight: 'bold',
        marginStart: 30
    },
    containerInput: {
        backgroundColor:'white',
        width:300,
        marginStart:30,
        marginBottom:10
    },
    containerDefault: {
        marginBottom:10,
        justifyContent:'space-between' 
    },
    inputDefault: {
        height: 50,
        backgroundColor: 'white',
        fontSize: 17,
        width:300,
        marginStart:30
    },
    containerButton: {
        marginTop: 30,
        marginStart: 20,
        marginEnd: 20,
        marginBottom: 2
    }
});

File: Flutter/Flutter - antigo/lib/pages/sat_pages/ativarSat.dart
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:flutter_gertec/services/globalValues.dart';
import 'package:flutter_gertec/services/operacaoSat.dart';
import 'package:flutter_gertec/services/retornoSat.dart';
import 'package:flutter_gertec/util/common_code.dart';
import 'package:flutter_gertec/widgets/widgetsgertec.dart';
import 'package:flutter_masked_text/flutter_masked_text.dart';

class PageAtivarSat extends StatefulWidget {
  @override
  _PageAtivarSatState createState() => _PageAtivarSatState();
}

class _PageAtivarSatState extends State<PageAtivarSat> {
  final cnpj = MaskedTextController(mask: "00.000.000/0000-00"); // CNPJ do contribuinte

  // Inicializa o código de ativação com um valor global, para o usuario não precisar ficar digitando
  final codigoAtivacao = TextEditingController(text: GlobalValues.codAtivarSat); // Codigo de ativação que vai ser utilizado

  final confirmacaoCodigo = TextEditingController(); // Confirmação do codigo

  CommonGertec commonGertec = new CommonGertec(); //* Classe que possui partes de código comuns em diversas telas

  // Função para validar os valores digitos pelo usuario e realizar a ativação do SAT
  void ativarSat() async {
    GlobalValues.codAtivarSat = codigoAtivacao.text; // Salva  o código de ativação para o usuario não precisar ficar digitando em todas as telas
    if (commonGertec.isCodigoValido(codigoAtivacao.text)) {
      if (codigoAtivacao.text == confirmacaoCodigo.text) {
        String cnpjNoMask = commonGertec.removeMaskCnpj(cnpj.text); // Remove a mascara do CNPJ para validar
        if (cnpjNoMask.length != 14) {
          WidgetsGertec.dialogoSat("Verifique o CNPJ digitado!", context: context);
        } else {
          //* Chama a função Invocar Operação Sat, que recebe como parâmetro a "operação invocada" e um "Map com as chaves e seus respectivos valores".
          RetornoSat retornoSat = await OperacaoSat.invocarOperacaoSat(
            // Passa como parâmetro um Map de argumentos de valores que deseja enviar
            args: {'funcao': 'AtivarSAT', 'random': Random().nextInt(999999), 'codigoAtivar': codigoAtivacao.text.toString(), 'cnpj': cnpjNoMask},
          );

          //? Caso deseje visualizar o retorno completo basta chamar [retornoSat.getResultadoCompleto]

          //* Está função [OperacaoSat.formataRetornoSat] recebe como parâmetro a operação realizada e um objeto do tipo RetornoSat
          //* Retorna uma String com os valores obtidos do retorno da Operação já formatados e prontos para serem exibidos na tela
          // Recomenda-se acessar a função e entender como ela funciona
          String retornoFormatado = OperacaoSat.formataRetornoSat(retornoSat: retornoSat);
          WidgetsGertec.dialogoSat(retornoFormatado, context: context);
        }
      } else {
        WidgetsGertec.dialogoSat("O Código de Ativação e a Confirmação do Código de Ativação não correspondem!", context: context);
      }
    } else {
      WidgetsGertec.dialogoSat("Código de Ativação deve ter entre 8 a 32 caracteres!", context: context);
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SingleChildScrollView(
        child: Container(
          color: Colors.white,
          padding: EdgeInsets.only(top: 30),
          height: MediaQuery.of(context).size.height,
          width: MediaQuery.of(context).size.width,
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.center,
            mainAxisSize: MainAxisSize.min,
            children: <Widget>[
              SizedBox(height: 30),
              WidgetsGertec.formField(cnpj, "CNPJ Contribuinte: ", textInputType: TextInputType.number),
              SizedBox(height: 20),
              WidgetsGertec.formField(codigoAtivacao, "Código de Ativação SAT: "),
              SizedBox(height: 20),
              WidgetsGertec.formField(confirmacaoCodigo, "Confirmação do Código: "),
              SizedBox(height: 20),
              SizedBox(
                width: 240,
                child: RaisedButton(
                  child: Text("ATIVAR"),
                  onPressed: () {
                    ativarSat();
                  },
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}
